third_party/xla_client/pjrt_computation_client.cc

```diff
@@ -131,22 +146,31 @@ PjRtComputationClient::ExecuteComputation(
   const PjRtComputation& pjrt_computation =
       dynamic_cast<const PjRtComputation&>(computation);
 
+  xla::PjRtDevice* pjrt_device = StringToPjRtDevice(device);
+  XLA_CHECK(pjrt_device->IsAddressable()) << pjrt_device->DebugString();
+
   std::vector<xla::PjRtBuffer*> buffers;
   buffers.reserve(arguments.size());
   for (auto& argument : arguments) {
     const PjRtData* pjrt_data = dynamic_cast<PjRtData*>(argument.get());
+
+    XLA_CHECK(pjrt_device == pjrt_data->buffer->device())
+        << pjrt_device->DebugString() << " vs "
+        << pjrt_data->buffer->device()->DebugString();
     buffers.push_back(pjrt_data->buffer.get());
   }
 
   xla::ExecuteOptions execute_options;
-  execute_options.untuple_result = true;
-  std::vector<std::vector<std::unique_ptr<xla::PjRtBuffer>>> results =
-      pjrt_computation.executable->Execute({buffers}, execute_options)
+  execute_options.untuple_result = options.explode_tuple;
+  execute_options.strict_shape_checking = false;
+  std::vector<std::unique_ptr<xla::PjRtBuffer>> results =
+      pjrt_computation.executable
+          ->ExecuteSharded(buffers, pjrt_device, execute_options)
```

Comment:
You've already explained to me, @will-cromar could you help me understand again why we are switching from Execute to ExecuteSharded here?

Reply:
Execute = execute one computation with a list of inputs on a list of devices (corresponds to [ComputationClient::ExecuteReplicated](https://github.com/pytorch/xla/blob/1ab841877eaaa7a0acd6774172bcc52315f04073/third_party/xla_client/computation_client.h#L217)]). Note how down below I had to fetch the 0th output vector before iterating through the outputs. If I left Execute here, every thread would execute every argument on every device, rather than executing the _i_-th set of arguments on the _i_-th device (or just crash).

ExecuteSharded = execute one computation with one set of inputs on one device, which is the same API as [ExecuteComputation](https://github.com/pytorch/xla/blob/1ab841877eaaa7a0acd6774172bcc52315f04073/third_party/xla_client/computation_client.h#L202)